### PR TITLE
Improve mrack logging

### PR DESCRIPTION
**Pull request by Tiboris**
refactor: Add dsp_name to logging messages to display context
    
    Add dsp_name to logging messages to display context
    for prioviders and transformers.
    Added some debug messages.
    
Aims to resolve https://github.com/neoave/mrack/issues/23 

Signed-off-by: Tibor Dudlák <tdudlak@redhat.com>

### CODE REVIEW

**Review by netoarmando — CHANGES_REQUESTED** [review]
This is breaking logging for some reason, I tried to execute this:
```bash
mrack --debug -c non-existent_prov_config.yaml up  non-existent_metadata.yaml
```
No output was returned.

**Comment by Tiboris** [comment]
> ```shell
> mrack --debug
> ```

yes you are correct when looking at this it should raise exception
https://github.com/neoave/mrack/blob/master/src/mrack/run.py#L194
and it raises it with the known_error when file is not found seems to be the:
<class 'mrack.errors.ConfigError'>
where it breaks, seems like exception_handler can not use logger with my changes.

**Comment by Tiboris** [comment]
@netoarmando I hope it is fixed, reason for that can be seen in file `__init__.py`
https://github.com/neoave/mrack/pull/45/files#diff-3b06fe18a2bbf59c91431116c264ffac811c2e81bf485eed06ca51536c7e896cR15
The decorator does not pass the right function name for the logger.

File: src/mrack/logging.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Logging interface."""
+import logging
+
+DEBUG = logging.DEBUG
```
Comment (netoarmando) [inline src/mrack/logging.py:18]: I guess you are defining this so other parts can import that from mrack.logging, right?

How come the linter is not complaining about this?

File: src/mrack/transformers/transformer.py [hunk 2]
```diff
@@ -72,3 +73,13 @@ def validate_host(self, host):
         provider = host.get("provider")
         if provider and provider not in providers.names:
             raise MetadataError(f"Error: Invalid host provider: {provider}")
+
+    def create_host_requirement(self, host):
+        """Create single input for provisioner."""
+        raise NotImplementedError()
+
+    def create_host_requirements(self):
```
Comment (netoarmando) [inline src/mrack/transformers/transformer.py:81]: nitpick: I would like to see these changes in another commit.

File: src/mrack/run.py [hunk 3]
```diff
@@ -198,7 +198,7 @@ def handle(*args, **kwargs):
             ProviderError,
             ApplicationError,
         ) as known_error:
-            logger.error(known_error)
+            logger.error(known_error, extra={"name_override": func.__name__})
```
Comment (netoarmando) [inline src/mrack/run.py:201]: In the end this was necessary. Can't we leave the logger configuration on `src/mrack/__init__.py`, so if any module don't need to set a context can just use the logger with no surprises? The reasoning is that we could call logger from anywhere, without worrying about it breaking in production.

File: src/mrack/transformers/transformer.py (hunk 2 above)
Comment (Tiboris) [reply src/mrack/transformers/transformer.py:81]: Will do.

File: src/mrack/run.py (hunk 3 above)
Comment (Tiboris) [reply src/mrack/run.py:201]: Hmmm this should not be needed.

File: src/mrack/logging.py (hunk 1 above)
Comment (Tiboris) [reply src/mrack/logging.py:18]: Really, dunno :D maybe because it is upper-case.

**Comment by pvoborni** [comment]
Imagine that you are a person who don't know the implementation of mrack. What is your feeling as the user when you are using mrack with this logging?

File: src/mrack/run.py (hunk 3 above)
Comment (Tiboris) [reply src/mrack/run.py]: I would use only mrack logging so we do not mix up the loggers and the log is consistent.

File: src/mrack/transformers/transformer.py (hunk 2 above)
Comment (Tiboris) [reply src/mrack/transformers/transformer.py]: Removed i will create another PR

**Comment by Tiboris** [comment]
> Imagine that you are a person who don't know the implementation of mrack. What is your feeling as the user when you are using mrack with this logging?

Could you please collaborate more on what do you have in mind? I am afraid that I do not understand what you are trying to point out. :(

**Comment by pvoborni** [comment]
> > Imagine that you are a person who don't know the implementation of mrack. What is your feeling as the user when you are using mrack with this logging?
> 
> Could you please collaborate more on what do you have in mind? I am afraid that I do not understand what you are trying to point out. :(

It's a question to you and Armando :) Don't want to point out anything yet.  Sort of a nudge to apply a perspective different from ours (developers).

**Comment by netoarmando** [comment]
> Imagine that you are a person who don't know the implementation of mrack. What is your feeling as the user when you are using mrack with this logging?

For me, the ideal scenario would be to use python standard logging call, BUT with the option to use our internal logging for context.. However, I do think that we can refer to the Zen of Python, which **might** mean to return to plain loggers with explicit strings (e.g.: `f"{provider}: message"` because it is simpler and explicit.

**Comment by pvoborni** [comment]
I actually never red Zen of Python, thanks for the reference. https://www.python.org/dev/peps/pep-0020/

In the spirit of it I'd add: "Experience for end users is more important than purity of implementation."

I'd say prefixing is desired only in transformers and providers and in debug level. In normal operations, in the rest of the app is not desired - almost no CLI apps do it. 

Going with simple logger could be the way. E.g. having a logging wrapper(s) with the prefixing in transformer and provider base class and using them in the logging calls in transformer/provider implementations.

**Comment by Tiboris** [comment]
From my point of view it is totally all-right to complicate logger little bit so it fulfil our needs and we use the same everywhere and as implementation looks like now, it is not really that complicated. I believe it is okay that project has its own logger.

File: src/mrack/transformers/transformer.py [hunk 4]
```diff
@@ -41,9 +42,13 @@ async def init(self, cfg, metadata):
         self._provider = providers.get(self._config_key)
         await self.init_provider()
 
+    def set_display_name(self, name):
```
Comment (netoarmando) [inline src/mrack/transformers/transformer.py:45]: It's a public property, I think this is not necessary,

File: src/mrack/providers/aws.py [hunk 5]
```diff
@@ -76,17 +76,22 @@ async def validate_hosts(self, hosts):
             req_img = req.get("image")
             if not req.get("meta_image") and req_img not in self.ami_ids:
                 raise ValidationError(
-                    f"{self.display_name} provider does not support "
-                    f"'{req_img}' image in provisioning config."
+                    f"{self.display_name}: Provider does not support "
```
Comment (netoarmando) [inline src/mrack/providers/aws.py:79]: `self.dsp_name`

Comment (Tiboris) [reply src/mrack/providers/aws.py:79]: right!

File: src/mrack/transformers/transformer.py (hunk 4 above)
Comment (Tiboris) [reply src/mrack/transformers/transformer.py:45]: allright

File: src/mrack/logging.py (hunk 1 above)
Comment (Tiboris) [reply src/mrack/logging.py]: This is not relevant anymore marking as resolved

**Review by netoarmando — APPROVED** [review]

**Pull request merged**